import necessary libs


In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

In [2]:
df=pd.read_csv('./dataset.csv')
df=df.drop('customerID',axis='columns')
df.head()

gender SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female            No     Yes         No       1           No   
1    Male            No      No         No      34          Yes   
2    Male            No      No         No       2          Yes   
3    Male            No      No         No      45           No   
4  Female            No      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges  TotalCharges  \
0              Yes           Electronic check           29.85         29.85   
1               No               Mailed check           56.95       1889.50   
2              Yes               Mailed check           53.85        108.15   
3               No  Bank transfer (automatic)           42.30       1840.75   
4              Yes           Electronic check           70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes

In [3]:
df.isnull().any()

gender              False
SeniorCitizen       False
Partner             False
Dependents          False
tenure              False
PhoneService        False
MultipleLines       False
InternetService     False
OnlineSecurity      False
OnlineBackup        False
DeviceProtection    False
TechSupport         False
StreamingTV         False
StreamingMovies     False
Contract            False
PaperlessBilling    False
PaymentMethod       False
MonthlyCharges      False
TotalCharges         True
Churn               False
dtype: bool

In [4]:
df['TotalCharges'] = df['TotalCharges'].replace(np.nan, df['TotalCharges'].mean())

In [5]:
numeric_columns = df.select_dtypes(exclude=['int64', 'float64']).columns
li=list(numeric_columns)
print("Numeric columns:",li )


Numeric columns: ['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'Churn']


In [6]:
lb=LabelEncoder()

for i in li:
    df[i]=lb.fit_transform(df[i])

df.head(3)

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0       0              0        1           0       1             0   
1       1              0        0           0      34             1   
2       1              0        0           0       2             1   

   MultipleLines  InternetService  OnlineSecurity  OnlineBackup  \
0              1                0               0             2   
1              0                0               2             0   
2              0                0               2             2   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                 0            0            0                0         0   
1                 2            0            0                0         1   
2                 0            0            0                0         0   

   PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges  Churn  
0                 1              2           29.85         29.85      0  
1                 0              3           56.95       1889.50      0  
2                 1              3           53.85        108.15      1

In [12]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import mutual_info_classif

def analyze_feature_importance(df, target_col):
    # Separate features and target
    X = df.drop(target_col, axis=1)
    y = df[target_col]
    
    # 1. Correlation with target
    correlations = abs(df.corr()[target_col]).drop(target_col)
    corr_scores = correlations.sort_values(ascending=False)
    
    # 2. Random Forest Importance
    rf = RandomForestClassifier(n_estimators=100, random_state=42)
    rf.fit(X, y)
    rf_importance = pd.Series(rf.feature_importances_, index=X.columns)
    rf_scores = rf_importance.sort_values(ascending=False)
    
    # 3. Mutual Information
    mi_scores = mutual_info_classif(X, y)
    mi_series = pd.Series(mi_scores, index=X.columns)
    mi_scores = mi_series.sort_values(ascending=False)
    
    # 4. Combine all scores
    # Normalize each score type
    normalized_scores = pd.DataFrame({
        'Correlation': (corr_scores - corr_scores.min()) / (corr_scores.max() - corr_scores.min()),
        'RandomForest': (rf_scores - rf_scores.min()) / (rf_scores.max() - rf_scores.min()),
        'MutualInfo': (mi_scores - mi_scores.min()) / (mi_scores.max() - mi_scores.min())
    })
    
    # Calculate average importance
    normalized_scores['Average_Importance'] = normalized_scores.mean(axis=1)
    final_ranking = normalized_scores.sort_values('Average_Importance', ascending=False)
    
    return final_ranking


importance_ranking = analyze_feature_importance(df, 'Churn')
print("Top 5 most important features:")
print(importance_ranking.head(5))

Top 5 most important features:
                Correlation  RandomForest  MutualInfo  Average_Importance
Contract           1.000000      0.410168    1.000000            0.803389
tenure             0.870704      0.836014    0.701123            0.802613
TotalCharges       0.491665      1.000000    0.443899            0.645188
MonthlyCharges     0.476022      0.967024    0.458610            0.633885
OnlineSecurity     0.723258      0.219954    0.644446            0.529219


### Highly Relavent features
| Feature | Correlation | RandomForest | MutualInfo | Average_Importance |
|---|---|---|---|---|
| tenure | 0.870704 | 0.836014 | 0.770066 | 0.825594 |
| Contract | 1.000000 | 0.410168 | 1.000000 | 0.803389 |
| MonthlyCharges | 0.476022 | 0.967024 | 0.499967 | 0.647671 |
| TotalCharges | 0.491665 | 1.000000 | 0.444237 | 0.645301 |
| OnlineSecurity | 0.723258 | 0.219954 | 0.742850 | 0.562021 |
| TechSupport | 0.705692 | 0.209430 | 0.668060 | 0.527727 |

In [8]:
# df=df.drop(['OnlineBackup','DeviceProtection','PaymentMethod','PaperlessBilling','SeniorCitizen','Partner','gender'],axis=1)
# df.head()

In [13]:
# Get the top 5 feature names from your importance ranking
top_5_features = importance_ranking.head(5).index.tolist()

# Create final dataset with only top 5 features + target column
final_columns = top_5_features + ['Churn']
final_df = df[final_columns]

print("Selected features:", top_5_features)
print("\nFinal dataset shape:", final_df.shape)

Selected features: ['Contract', 'tenure', 'TotalCharges', 'MonthlyCharges', 'OnlineSecurity']

Final dataset shape: (7043, 6)


In [14]:
final_df.to_csv('final_dataset.csv', index=False)